CryptoAnalysis/src/main/java/crypto/providerdetection/ProviderDetection.java

```diff
@@ -0,0 +1,513 @@
+		String value = providerValue.toString();
+		for(Unit unit : body.getUnits()) {
+			if(unit instanceof JIfStmt) {
+				JIfStmt ifStatement = (JIfStmt) unit;
+				if(ifStatement.toString().contains(value)) {
+					throw new RuntimeException("The provider parameter must be passed directly to the"
+							+ " getInstance() method call, and not through IF-ELSE statements or"
+							+ " TERNARY operators.");
+				}
+			}
+		}
+		return;
+	}
+	
+	
+	/**
+	 * This method checks if the provider detected has only one allocation site
+	 * and it is not flowing through SWITCH statements, because otherwise the 
+	 * provider can not be correctly detected through the use of static analysis
+	 * 
+	 * @param providerValue
+	 *            
+	 * @param body
+	 *            - i.e. the ActiveBody
+	 *            
+	 */
+	private void checkSwitchStmt(Value providerValue, Body body) {
+		String value = providerValue.toString();
+		for(Unit unit : body.getUnits()) {
+			if(unit instanceof TableSwitchStmt) {
+				TableSwitchStmt switchStatement = (TableSwitchStmt) unit;
+				if(switchStatement.toString().contains(value)) {
+					throw new RuntimeException("The provider parameter must be passed directly to the"
+							+ " getInstance() method call, and not through SWITCH statements.");
+				}
+			}
+		}
+		return;
+	}
+	
+	
+	/**
+	 * This method is used to check if the CryptSL rule for the detected provider exists
+	 * 
+	 * @param provider
+	 *            - i.e. BC
+	 * @param declaringClassName
+	 *            - i.e. MessageDigest
+	 */
+	private boolean ruleExists(String provider, String declaringClassName) {
+		boolean ruleExists = false;
+		String rule = declaringClassName;
+		
+		File rulesDirectory = new File(defaultRulesDirectory+File.separator+provider);
+		if(rulesDirectory.exists()) {
+			File[] listRulesDirectoryFiles = rulesDirectory.listFiles();
+			for (File file : listRulesDirectoryFiles) {
+				if (file != null && file.getAbsolutePath().endsWith(rule+".cryptslbin")) {
+					ruleExists = true;
+					break;
+				}
+			}
+		}
+		
+		return ruleExists;
+	}
+	
+	
+	/**
+	 * This method is used to choose the CryptSL rules from the detected Provider
+	 * 
+	 * @param rules
+	 *            
+	 * @param provider
+	 *            - i.e. BC
+	 * @param declaringClassName
+	 * 			  - i.e. MessageDigest
+	 */
+	private List<CryptSLRule> chooseRules(List<CryptSLRule> rules, String provider, String declaringClassName) {
+		
+		String newRulesDirectory = defaultRulesDirectory+File.separator+provider;
+		
+		// Forms a list of all the new CryptSL rules in the detected provider's directory.
+		// This list contains only String elements and it holds only the rule's names, i.e Cipher, MessageDigest, etc
+		List<String> newRules = new ArrayList<String>();
+		File[] files = new File(newRulesDirectory).listFiles();
+		for (File file : files) {
+		    if (file.isFile() && file.getName().endsWith(".cryptslbin")) {
```

Comment:
Please introduce a Constant for `".crytslbin"` and also find any occurences in your classes *and* the whole project and replace them by the new constant.